### PR TITLE
[misc] add helper script to update macros

### PR DESCRIPTION
**Pull request by gjc13**
Add helper script to replace changed config macro names.
Also force configure to fail upon unrecognized options.

### CODE REVIEW

File: tools/macro_update/replace.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+#  Usage:
+#  replaces.sh OPENTHREAD_CONFIG_HEADER_FILE
+
+sed -i 's/OPENTHREAD_CONFIG_ENABLE_TIME_SYNC/OPENTHREAD_CONFIG_TIME_SYNC_ENABLE/g' $1
```
Comment (bukepo) [inline tools/macro_update/replace.sh]: match `\bOPENTHREAD_CONFIG_ENABLE_TIME_SYNC\b`?

File: tools/macro_update/replace.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
+#!/bin/bash
+#
+#  Copyright (c) 2018, The OpenThread Authors.
```
Comment (bukepo) [inline tools/macro_update/replace.sh]: 2019

File: tools/macro_update/replace.sh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+#  Usage:
+#  replaces.sh OPENTHREAD_CONFIG_HEADER_FILE
+
+sed -i 's/OPENTHREAD_CONFIG_ENABLE_TIME_SYNC/OPENTHREAD_CONFIG_TIME_SYNC_ENABLE/g' $1
+sed -i 's/OPENTHREAD_ENABLE_BORDER_ROUTER/OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE/g' $1
```
Comment (bukepo) [inline tools/macro_update/replace.sh]: quote `$1`

File: tools/macro_update/replace.sh [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+sed -i 's/OPENTHREAD_ENABLE_SERVICE/OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE/g' $1
```
Comment (bukepo) [inline tools/macro_update/replace.sh]: replace in a single sed command by using multiple `-e` for each command?

File: tools/macro_update/replace.sh [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+#  Usage:
+#  replaces.sh OPENTHREAD_CONFIG_HEADER_FILE
```
Comment (bukepo) [inline tools/macro_update/replace.sh]: the file is named `replace.sh`, but here is `replaces.sh`.

How about we put it into `script/update-macros.sh`? And provide an example to use together with `find`?

File: tools/macro_update/replace.sh [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+          's/\bOPENTHREAD_CONFIG_MAX_ROUTERS\b/OPENTHREAD_CONFIG_MLE_MAX_ROUTERS/g'  \
+          's/\bOPENTHREAD_CONFIG_DEFAULT_CHILD_TIMEOUT\b/OPENTHREAD_CONFIG_MLE_CHILD_TIMEOUT_DEFAULT/g'  \
+          's/\bOPENTHREAD_CONFIG_IP_ADDRS_PER_CHILD\b/OPENTHREAD_CONFIG_MLE_IP_ADDRS_PER_CHILD/g'  \
+          's/\bOPENTHREAD_CONFIG_SUPERVISION_CHECK_TIMEOUT\b/OPENTHREAD_CONFIG_CHILD_SUPERVISION_CHECK_TIMEOUT/g'  \
+          's/\bOPENTHREAD_CONFIG_NUM_SLAAC_ADDRESSES\b/OPENTHREAD_CONFIG_IP6_SLAAC_NUM_ADDRESSES/g'  \
+          's/\bOPENTHREAD_ENABLE_NCP_UART\b/OPENTHREAD_CONFIG_NCP_UART_ENABLE/g'  \
+          's/\bOPENTHREAD_ENABLE_NCP_SPI\b/OPENTHREAD_CONFIG_NCP_SPI_ENABLE/g'  \
+          's/\bOPENTHREAD_ENABLE_SPINEL_VENDOR_SUPPORT\b/OPENTHREAD_ENABLE_NCP_VENDOR_HOOK/g'  \
+          's/\bOPENTHREAD_ENABLE_NCP_SPINEL_ENCRYPTER\b/OPENTHREAD_CONFIG_NCP_SPI_ENABLENEL_ENCRYPTER/g'  \
+          's/\bOPENTHREAD_ENABLE_PLATFORM_NETIF\b/OPENTHREAD_CONFIG_PLATFORM_NETIF_ENABLE/g'  "${1}"
```
Comment (bukepo) [inline tools/macro_update/replace.sh]: ` "${1}"` -> `"$1"`

File: tools/macro_update/replace.sh [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+#  Usage:
+#  replaces.sh OPENTHREAD_CONFIG_HEADER_FILE
+
+sed -i -e 's/\bOPENTHREAD_CONFIG_ENABLE_TIME_SYNC\b/OPENTHREAD_CONFIG_TIME_SYNC_ENABLE/g'  \
+          's/\bOPENTHREAD_ENABLE_BORDER_ROUTER\b/OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE/g'  \
```
Comment (bukepo) [inline tools/macro_update/replace.sh]: add `-e` and sort?

File: script/replace_macros.sh [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+#  Usage:
+#  replace_macros.sh OPENTHREAD_CONFIG_HEADER_FILE
+
+sed -i \
```
Comment (abtink) [inline script/replace_macros.sh]: While I agree having such scripts is very helpful in transitioning projects from older config definitions to new model, I am not sure if it makes sense to include this in OpenThread GitHub master branch. OT master does not support any of the older config definitions. I think including such a script can be confusing to a new developer/user of OT exploring the code.

In the history of development of OpenThread, we have had changes to the public OT APIs. Two larger ones that I recall:

- [PR #1764](https://github.com/openthread/openthread/pull/1764) where `otError` enumeration names  were changed from `kThreadError<ErrorInCamelCase>` model to `OT_ERROR_<ErrorInUpperCase>` model (the type itself was renamed `ThreadError` -> `otError`).
- [PR #1173 ](https://github.com/openthread/openthread/pull/1173) where many of public APIs and definitions were renamed to follow a consistent coding/naming convention. 

In such cases, we used scripts and tools to update all internal projects and convert the old naming model to new one, but we did not merge/include the scripts in the OpenThread master (github master branch). 

Thoughts?

Comment (gjc13) [reply script/replace_macros.sh]: Maybe we can put this script on gist and provide a link in openthread-core-config-check.h?

Comment (bukepo) [reply script/replace_macros.sh]: @abtink I think this tool is mainly for OpenThread users, who might still use a very old version of OpenThread(I used to find this fact in some questions raised in openthread-users group).  And once they want to upgrade to the latest OpenThread, they will find it difficult. That's why I think it makes sense to provide and maintain a script to migrate old code to latest code.

Although we don't support any of the older config definitions, we still provided `openthread-config-default-check.h` to check those old definitions are not defined. That's helpful to report a problem. But If I am an OpenThread user, I would wish to have a tool to help me solve the problem if possible. I was just asked to help migrate https://github.com/openthread/silk, I find this tool is very helpful.

So I tend to maintain this script, and every time we make a change, we should update this script to migrate the code. This also helps us to do GitHub sync. Thoughts?

Comment (abtink) [reply script/replace_macros.sh]: @bukepo, @gjc13 I 100% agree that such script would be useful. Only considering where to save/store such scripts (master or ...?)
On some internal project/repos, we used a `scripts` branch that included useful scripts for that repo. I think `wpantund` adopted same model [(scripts branch)](https://github.com/openthread/wpantund/tree/scripts).
I also like the idea of saving it on a gist.

Comment (gjc13) [reply script/replace_macros.sh]: @jwhui Could you shed some lights on this issue?

**Pull request closed without merging**